ontology/ontology.go

```diff
@@ -1,2 +1,264 @@
+	var answer = make(map[string]float64)
+	var totalBases, currLen int = 0, 0
+	var currGene string
+
+	for i := range filledSpace {
+		currLen = filledSpace[i].ChromEnd - filledSpace[i].ChromStart
+		currGene = strings.ToUpper(filledSpace[i].Name)
+		answerCounts[currGene] += currLen
+		totalBases += currLen
+	}
+
+	for i := range answerCounts {
+		answer[i] = float64(answerCounts[i]) / float64(totalBases)
+	}
+
+	return answer
+}
+
+// termProportionOfGenome returns a map[string]float64, where keys are the names of a gene ontology term and values are the
+// proportion of the genome covered by that term.
+func termProportionOfGenome(ontologies map[string]*Ontology, geneProportions map[string]float64) map[string]float64 {
+	var answer = make(map[string]float64)
+	var j int
+	for i := range ontologies {
+		answer[i] = 0
+		for j = range ontologies[i].Genes {
+			answer[i] += geneProportions[strings.ToUpper(ontologies[i].Genes[j])]
+		}
+	}
+	return answer
+}
+
+// ThreeDGreat uses all inputs to determine which areas of the genome, based on 3D genome contacts, are associated with
+// a gene, and therefore it's function, and will output the enrichment score for a given GO term.
+// This function takes in a query file as a bed struct, a chrom.sizes file, a map of genes in gtf form, a contact file in
+// bedpe form, an annotation file in gaf form and a map that takes GO term ID's and relates them to other GO term features
+// in obo format, and finally it takes a string which will write out all regions of the genome with their assigned
+// genes in the name column and their assigned ontologies in the annotation field.
+func ThreeDGreat(queries []bed.Bed, chromSizes map[string]chromInfo.ChromInfo, geneFile string, contacts []bedpe.BedPe, annotations []gaf.Gaf, oboMap map[string]*obo.Obo, out3dOntology string, geneEnrichments bool, termEnrichments bool) {
+	var err error
+	var filledSpace, tssBed []bed.Bed
+	var filledSpaceIntervals []interval.Interval
+	var queryOverlaps []interval.Interval
+	var kCache = make(map[string]int) //mapping Go Term Names to number of query elements that overlap. This is 'k' in the binomial test.
+	var n = len(queries)              // this stores the number of query elements, which is the number of trials in the binomial test.
+	var currOverlapGene string
+	var currOntologyIndex int
+	var currOntologyName string
+	name := strings.TrimSuffix(out3dOntology, ".bed")
+	var ontologiesForCurrGene []*Ontology
+
+	geneString := strings.Split(geneFile, ".")
+	if geneString[len(geneString)-1] != "bed" {
+		genes := gtf.Read(geneFile)
+		tssBed = gtf.GenesToTssBed(genes, chromSizes, true)
+	} else {
+		tssBed = bed.Read(geneFile)
+	}
+	bed.SortByCoord(tssBed)
+	filledSpace = Fill3dSpace(contacts, tssBed, chromSizes)
+	ontologies := OboToOntology(oboMap)
+	GeneAssignmentsFromGaf(annotations, ontologies)
+	geneOntologies := GenesToOntologies(ontologies)
+
+	//TODO: filtering out genes without ontology annotations? as an option?
```

Comment:
They are added to the map, but they key to nil pointers, Which I think right now is the best case, since so many different genes files will inevitably contain genes without ontologies, and later we examine gene space in the genome and then term space, and I have a filter later that checks to make sure a GO term actually is included in the genome space before doing any math for it